File: russian_dolls.py
'''
Time Complexity : O(nlogn) in worst case run Binary search for all elements in the nums list
Space Complexity : O(n) store elements in arr
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
Approach: First step is to sort the envelopes array in increasing order of first index and decreasing order of last index.
Next is similar as longest increasing subsequence problem.
'''

class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        if not envelopes:
            return 0
        envelopes.sort(key=lambda x: (x[0], -x[1]))
        length = len(envelopes)
        arr = []
        arr.append(envelopes[0][1])
        for i in range(1, length):
            if envelopes[i][1] > arr[-1]:
                arr.append(envelopes[i][1])
            else:
                position = self.binarySearch(arr, 0, len(arr)-1, envelopes[i][1])
                arr[position] = envelopes[i][1]
        return len(arr)

    def binarySearch(self, arr, low, high, target):
        while low <= high:
            mid = (low+high)//2
            if arr[mid] == target:
                return mid
            if arr[mid] > target:
                high = mid-1
            else:
                low = mid+1
        return low

File: russiandolls.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
sorted the given list(either acc to width or height) used Binary search method to place each element in the list at its position
class Solution {
    public int maxEnvelopes(int[][] envelopes) {
        Arrays.sort(envelopes,(a,b)-> {
            if(a[0]==b[0]){
                return b[1]-a[1];
            }
            else return a[0]-b[0];
            });
        int n=envelopes.length;
        int []arr = new int[n];
        arr[0] = envelopes[0][1];
        int len=1;
        for(int i=1;i<n;i++){
            if(envelopes[i][1] > arr[len-1]){
                arr[len] = envelopes[i][1];
                len++;
            }
            else{
                int bsindex = Bsearch(arr,envelopes[i][1],0,len-1);
                arr[bsindex] = envelopes[i][1];
            }
            
        }
        return len;
    }
    private int Bsearch(int [] arr,int target,int low,int high){
        int mid;
        while(low<=high){
            mid = low+(high-low)/2;
            if(arr[mid] == target) return mid;
            else if(arr[mid] < target) low= mid+1;
            else high = mid-1;
        }
        return low;
    }
}